api/cloud/oc_cloud_endpoint.c

```diff
@@ -0,0 +1,285 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#include "oc_cloud_endpoint_internal.h"
+#include "port/oc_log_internal.h"
+#include "util/oc_list.h"
+#include "util/oc_memb.h"
+
+#include <string.h>
+
+// max two endpoint addresses per device with static allocation
+OC_MEMB(g_cloud_endpoint_s, oc_cloud_endpoint_t,
+        OC_CLOUD_MAX_ENDPOINT_ADDRESSES);
+
+static oc_cloud_endpoint_t *
+cloud_endpoint_item_allocate_and_add(oc_cloud_endpoints_t *ce,
+                                     oc_string_view_t uri)
+{
+  oc_cloud_endpoint_t *cei = oc_memb_alloc(&g_cloud_endpoint_s);
+  if (cei == NULL) {
+    return NULL;
+  }
+  oc_new_string(&cei->uri, uri.data, uri.length);
+  if (oc_string(cei->uri) == NULL) {
+    oc_memb_free(&g_cloud_endpoint_s, cei);
+    return NULL;
+  }
+
+  oc_list_add(ce->endpoints, cei);
+  return cei;
+}
+
+static void
+cloud_endpoint_item_free(oc_cloud_endpoint_t *cei)
+{
+  oc_free_string(&cei->uri);
+  oc_memb_free(&g_cloud_endpoint_s, cei);
+}
+
+const oc_string_t *
+oc_cloud_endpoint_uri(const oc_cloud_endpoint_t *endpoint)
+{
+  return &endpoint->uri;
+}
+
+bool
+oc_cloud_endpoints_init(oc_cloud_endpoints_t *ce, oc_string_view_t default_uri)
+{
+  memset(ce, 0, sizeof(oc_cloud_endpoints_t));
+  OC_LIST_STRUCT_INIT(ce, endpoints);
+
+  if (default_uri.length > 0) {
+    const oc_cloud_endpoint_t *cei =
+      cloud_endpoint_item_allocate_and_add(ce, default_uri);
+    if (cei == NULL) {
+      return false;
+    }
+    ce->selected = cei;
+  }
+  return true;
+}
+
+void
+oc_cloud_endpoints_deinit(oc_cloud_endpoints_t *ce)
+{
+  if (ce->endpoints == NULL) {
+    return;
+  }
+  oc_cloud_endpoint_t *cei = oc_list_pop(ce->endpoints);
+  while (cei != NULL) {
+    cloud_endpoint_item_free(cei);
+    cei = oc_list_pop(ce->endpoints);
+  }
+}
+
+bool
+oc_cloud_endpoints_reinit(oc_cloud_endpoints_t *ce,
+                          oc_string_view_t default_uri)
+{
+  oc_cloud_endpoints_deinit(ce);
+  return oc_cloud_endpoints_init(ce, default_uri);
+}
+
+size_t
+oc_cloud_endpoints_size(const oc_cloud_endpoints_t *ce)
+{
+  return ce->endpoints == NULL ? 0 : oc_list_length(ce->endpoints);
+}
+
+bool
+oc_cloud_endpoints_is_empty(const oc_cloud_endpoints_t *ce)
+{
+  return oc_cloud_endpoints_size(ce) == 0;
+}
+
+void
+oc_cloud_endpoints_iterate(const oc_cloud_endpoints_t *ce,
+                           oc_cloud_endpoints_iterate_fn_t fn, void *data)
+{
+  if (ce->endpoints == NULL) {
+    return;
+  }
+  oc_cloud_endpoint_t *cei = (oc_cloud_endpoint_t *)oc_list_head(ce->endpoints);
+  while (cei != NULL) {
+    if (!fn(cei, data)) {
```

Comment:
in `fn` user can delete `cei` so we need to guard this.

Reply:
added a comment to documentation